fog/sig/src/public_address.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright 2018-2021 The MobileCoin Foundation
+
+//! This module provides the implementation of the all-in-one verifier for
+//! public addresses.
+
+use crate::{Error, Verifier};
+use core::convert::TryInto;
+use mc_account_keys::PublicAddress;
+use mc_crypto_keys::Ed25519Signature;
+use mc_crypto_x509_utils::{
+    PublicKeyType, X509CertificateChain, X509CertificateIter, X509KeyExtrator,
+};
+use mc_fog_sig_authority::Verifier as AuthorityVerifier;
+use mc_fog_sig_report::Verifier as ReportVerifier;
+use mc_fog_types::ReportResponse;
+use signature::{Error as SignatureError, Signature};
+use x509_signature::X509Certificate;
+
+impl Verifier for PublicAddress {
+    type ReportSigError = SignatureError;
+
+    fn verify_fog_sig(
+        &self,
+        report_response: &ReportResponse,
+    ) -> Result<(), Error<<Self as AuthorityVerifier>::Error, Self::ReportSigError>> {
+        // Vec<Vec<u8>> -> Vec<&[u8]>
+        // Vec<&[u8]> -> Vec<X509Certificate>
+        let certs = X509CertificateIter::from(
+            report_response
+                .chain
+                .iter()
+                .map(|der| der.as_slice())
+                .collect::<Vec<&[u8]>>(),
+        )
+        .collect::<Vec<X509Certificate>>();
+
+        // Verify the authority signature
+        let authority_sig = self
+            .fog_authority_sig()
+            .ok_or(Error::NoSignature)?
+            .try_into()?;
+
+        self.verify_authority(&certs[0].subject_public_key_info().spki(), &authority_sig)
```

Comment:
If the report contained a bunch of invalid signatures (or just the first one was invalid) then `certs` would be empty and we'll panic.